lib/matplotlib/__init__.py

```diff
@@ -851,41 +852,51 @@ def __getitem__(self, key):
             version, alt_key, alt_val, inverse_alt = _deprecated_map[key]
             cbook.warn_deprecated(
                 version, key, obj_type="rcparam", alternative=alt_key)
-            return inverse_alt(dict.__getitem__(self, alt_key))
+            return inverse_alt(self._data[alt_key])
 
         elif key in _deprecated_ignore_map:
             version, alt_key = _deprecated_ignore_map[key]
             cbook.warn_deprecated(
                 version, key, obj_type="rcparam", alternative=alt_key)
-            return dict.__getitem__(self, alt_key) if alt_key else None
+            return self._data[alt_key] if alt_key else None
 
         elif key == 'examples.directory':
             cbook.warn_deprecated(
                 "3.0", "{} is deprecated; in the future, examples will be "
                 "found relative to the 'datapath' directory.".format(key))
 
         elif key == "backend":
-            val = dict.__getitem__(self, key)
+            val = self._data[key]
             if val is rcsetup._auto_backend_sentinel:
                 from matplotlib import pyplot as plt
                 plt.switch_backend(rcsetup._auto_backend_sentinel)
 
-        return dict.__getitem__(self, key)
+        return self._data[key]
 
     def __repr__(self):
         class_name = self.__class__.__name__
         indent = len(class_name) + 1
-        repr_split = pprint.pformat(dict(self), indent=1,
+        repr_split = pprint.pformat(self._data, indent=1,
                                     width=80 - indent).split('\n')
         repr_indented = ('\n' + ' ' * indent).join(repr_split)
         return '{}({})'.format(class_name, repr_indented)
 
     def __str__(self):
-        return '\n'.join(map('{0[0]}: {0[1]}'.format, sorted(self.items())))
+        return '\n'.join(map('{0[0]}: {0[1]}'.format,
+                             sorted(self._data.items())))
 
     def __iter__(self):
         """Yield sorted list of keys."""
-        yield from sorted(dict.__iter__(self))
+        yield from sorted(self._data)
+
+    def __len__(self):
+        return len(self._data)
+
+    def __delitem__(self, key):
```

Comment:
I actually don't think we should allow del'ing items from the rcParams as that has no sense.  It may not have been worth it to prevent that with the old implementation, but now it's just a matter of replacing the implementation by a raise TypeError.

Reply:
Yes, you're right. The current code prevents it, because we did not reimplement the `MutableMapping.__delitem__`. Before we were just inheriting from `dict`, so del'ing should have been possible. Inheriting `MutableMapping` was introduced within a large commit #7549. I dare assuming that both behaviors were incidental and nobody really thought about it.


I'm extending the RcParams in my private code to be able to style and style-context some of my plot elements in specialized plots. While I'm currently not del'ing anything I could imagine that there's a use-case for that. Maybe we should prevent del'ing only for values that are in the defaults.

But as a first step it's ok to stay as restrictive as we currently are.

Reply:
Talking about custom extensions to rcParams (none of which are really using the public API I guess), I realized that this PR would break https://github.com/anntzer/mplcairo#antialiasing where I use `dict.__setitem__` to bypass the validation code on input (to set lines.antialiased to one of the many cairo antialiasing modes, whereas the validator just accepts True/False).
I guess that's related to your point in https://github.com/matplotlib/matplotlib/pull/12577#issuecomment-431620474...

Reply:
I was hoping nobody would mess with the implementation details of `RcParams` in such a way except our own code. Ok, then we have to take it a step slower, define an official API for that and only later migrate to the `MutableMapping`.

By the way, `What was your reason for introducing the multiple inheritance in the first place? IMO deriving from `dict` is not good but still far better than deriving from `dict` and `MutableMapping`.

Reply:
> I was hoping nobody would mess with the implementation details of RcParams in such a way except our own code. Ok, then we have to take it a step slower, define an official API for that and only later migrate to the MutableMapping.

To be fully consistent with my own general approach wrt backcompat ("we should consider whether it'll actually matter to enough people that we reject a backcompat break"): I'm not going to throw a fit if the API changes here, and having a "more official" way to bypass the validator would be better anyways.

> By the way, What was your reason for introducing the multiple inheritance in the first place? IMO deriving from dict is not good but still far better than deriving from dict and MutableMapping.

When I wrote it, it seemed to be a good way to provide update() cheaply.
In fact, looking at it again:
- I guess you can still make it just inherit from `dict`, keeping most of the old code, and then do `update = MutableMapping.update` (thus easily stealing the implementation from MutableMapping without pulling in the whole multiple inheritance problem, and not breaking my approach in mplcairo :p);
- I think the fact that len(rcParams) == 0, i.e. that calls `MutableMapping.__len__`, is a CPython bug: https://bugs.python.org/issue35063

Reply:
LGTM